### DISCUSSION TITLE
Support profile definition under specific path

### DISCUSSION THREAD

**Discussion by skwde** [quarto-dev/quarto-cli#10192] [Feature Requests]
### Description

Having multiple profiles for a project can make the base directory of the project fairly crowded.
Besides that it might be

Being able to store profile specific stuff under a specfic path would solve that.

Borrowing the example at https://github.com/quarto-dev/quarto-cli/discussions/9520 for a debug profile instead of only allowing a flat structure like

```
_quarto-debug.yml
_quarto.yml
debug.css
talk.qmd
```

something like

```
profiles/debug/_quarto-debug.yml
profiles/debug/debug.css
_quarto.yml
talk.qmd
```

**Comment by mcanouil** [comment]
`_quarto.yml` files are by design at the root as they define the project.
That's unlikely to change, so if there is a change, it won't contain the prefix profiles currently have.

Also this would create more confusion with the `_metadata.yml` at the directory level (which for now does not support profile but will at some point).

How many profiles do you have that it messes your project? For instance Quarto' documentation has four and that is really not the source of a mess or "crowded" https://github.com/quarto-dev/quarto-web

Note that moving the profiles into directory does not change much the crowded fact unless you really have a lot of profiles.

Anyhow, thanks for the suggestion.
As this change might require deep changes and further design discussion, I can already tell you that if it happens it is not in a short time frame.

PS: the idea of a "profile" directory has the advantage of making very clear the project is using profiles.

**Reply by skwde** [reply]
Well I don't have that much profiles to be honest (I also only started using quarto a while back) but I just dislike the idea of having various things which 'might' be needed in the root of the project.

Possibly related to this "feature request" is support for project independent profiles, meaning that they can be located somewhere else to be shared by several projects.
Sticking to the debug example mentioned above, this is something helpful to all revealjs projects.

I any case, thanks for your explanation.

**Reply by mcanouil** [reply]
> Possibly related to this "feature request" is support for project independent profiles, meaning that they can be located somewhere else to be shared by several projects.

This is not going to happen.
The whole philosophy of Quarto projects is to make those projects self-contained
and reproducible.
Having files elsewhere would basically break that while also make it harder to debug things, i.e., are you sure you don't have a "global configuration"? sort of mess.

If you want to do things like that, then you are basically on your own esp if you go to the hard links path.
Alternatively, extensions are one way to "share" a setup in multiple projects (not across).

**Reply by skwde** [reply]
I am not quite sure what you mean by "global configuration" mess.

I basically try to find a way to work with quarto while not having to copy multiple files to several different projects (and repositories). 

Think of creating lectures, where each lecture is its own project and contains slides (revealjs -> pdf), exercises (pdf), solutions (pdf). Updating the "default" files in all repositories is a real mess and inconvenient as hell (I currently use rsync to update the individual projects... I am desperately waiting for https://github.com/quarto-dev/quarto-cli/issues/9515).

I am happy to stick to the "recommendation" & "intended" use of quarto but currently (at least to my understanding) it is not really possible to do this while relying to quarto extensions / templates.

What is the recommended way to share and update the something like the debug profile across multiple projects?
What is the recommended way to share and update project setup related files / folders (.gitignore, .devcontainer, linter config files, CI/CD setup, ...) across multiple projects?

**Reply by cderv** [reply]
> What is the recommended way to share and update the something like the debug profile across multiple projects?

We recently added support for metadata extensions for project. This is a way to share some configurations for project without creating a custom project type.  See https://quarto.org/docs/extensions/metadata.html
It is possible this serve well your profile use case and defining profile is a project configuration 

> What is the recommended way to share and update project setup related files / folders (.gitignore, .devcontainer, linter config files, CI/CD setup, ...) across multiple projects?

This is not quarto specific. Though if you are always referring to usage of those in Quarto context, maybe what you are looking for it a Project template. We call that Starter Template: https://quarto.org/docs/extensions/starter-templates.html

It is a way to easily share template with common content, usuful when always used in its own project. 

For more specific usage of syncing resources across project, especially when custom and not directly Quarto related, leveraging Project Scripts (https://quarto.org/docs/projects/scripts.html#pre-and-post-render) is a good way. A pre render script to check if all required resources exist and up to date with there latest version shared in a central place. 

Hope this helps

**Reply by skwde** [reply]
Thanks for your detailed response.

---

> This is not quarto specific. Though if you are always referring to usage of those in Quarto context, maybe what you are looking for it a Project template. We call that Starter Template: https://quarto.org/docs/extensions/starter-templates.html

To my understanding Starter Templates do not support hidden files (https://quarto.org/docs/extensions/starter-templates.html#creating-a-template) and an old discussion (https://github.com/quarto-dev/quarto-cli/discussions/9351) resulting also in https://github.com/quarto-dev/quarto-cli/issues/9515.

---

I think the current way to go (until project templates / starter templates support hidden files) is to use a metadata extension which does the setup / update using a pre-render script.
